proposals/0000-gadt-syntax.rst

```diff
@@ -0,0 +1,305 @@
+   #. ``->`` as a separator, which is special syntax rather than
+      the ``(->)`` type constructor;
+   #. the result type ``T``;
+
+3. A prefix-style GADT constructor signature has zero or more of the following
+   parts in a free order:
+
+   #. type variable telescope ``forall tvs.`` or ``forall tvs ->``;
+   #. constraint context ``ctx =>``;
+   #. parameter ``a ->``, ``a %p ->``, or ``a ->.``, possibly with a strictness
+      annotation;
+
+   It then ends with the result type ``T``.
+
+
+Proposed Change Specification
+-----------------------------
+
+Existing non-terminals
+~~~~~~~~~~~~~~~~~~~~~~
+
+In this section we describe the non-terminals, the definition of which is out
+of scope of this proposal (they are already present in the GHC grammar)
+
+1. ``atype`` - syntactically atomic types, e.g. ``T``, ``42``, ``x``, ``( ... )``, ``[ ... ]``
+2. ``ftype`` - application chain consisting of ``atype``, e.g. ``T @a b``
+3. ``btype`` - infix operator chain consisting of ``ftype``, e.g. ``T @a b + Q 42``
+4. ``forall_telescope`` - either ``forall tvs.`` or ``forall tvs ->``
+5. ``sig_vars`` - comma-separated field names, e.g. ``fld_1, fld_2, fld_n``, n >= 1
+6. ``mult`` - multiplicity annotation, e.g. ``%1``
+
+Proposed non-terminals
+~~~~~~~~~~~~~~~~~~~~~~
+
+Constructor declarations::
+
+  gadt_con ::= conids '::' gadt_prefix_sig
+            |  conids '::' gadt_record_sig
+
+  conids ::= conid
+          |  conid ',' conids
+
+
+Prefix form::
+
+  gadt_prefix_sig ::= btype
+                   |  quantifier gadt_prefix_sig
+
+  quantifier ::= forall_telescope
+              |  btype '=>'
+              |  fieldtype '->'
+              |  fieldtype '->.'
+              |  fieldtype mult '->'
+
+Record syntax::
+
+  gadt_record_sig ::= opt_forall opt_ctx '{' fielddecls '}' '->' btype
+
+  opt_forall ::= <empty>
+              |  forall_telescope
+
+  opt_ctx ::= <empty>
+           |  btype '=>'
+
+  fielddecls ::= fielddecl
+              |  fielddecl ',' fielddecl
+
+  fielddecl ::= sig_vars '::' fieldtype
+
+Field types::
+
+  fieldtype ::= opt_unpack btype
+              | opt_unpack strictness_sigil atype
+
+  opt_unpack ::= <empty>
+              |  '{-# UNPACK #-}'
+              |  '{-# NOUNPACK #-}'
+
+  strictness_sigil ::= '!'
+                    |  '~'
+
+Side conditions:
+
+* In ``gadt_record_sig``, the ``opt_forall`` must be of the ``forall tvs.`` form, not ``forall tvs ->``.
+* In ``strictness_sigil``, the ``!`` and ``~`` are assumed to be prefix occurrences.
+
+
+Effect and Interactions
+-----------------------
+
+Since we drop support for parentheses, the following programs will be all
+rejected by the new grammar::
+
+  data T where
+    T1 :: (Int -> T)
+    T2 :: (forall a. a -> T)
+    T3 :: forall a. (a -> T)
+    T4 :: forall a. (a -> (T))
+
+This makes it possible to avoid conflicts between ``btype`` and
+``gadt_prefix_sig`` in the LALR grammar.
+
+On the other hand, we now allow nested ``forall`` and contexts::
```

Comment:
This point comes up rather late and almost by-the-way.  But it will have a pretty major effect on the implementation.  It may also make it harder to specify pattern matching (see the [Type variables in patterns](https://www.microsoft.com/en-us/research/publication/type-variables-patterns/) paper).

I can see the motivation, but I'd like the proposal to draw attention to this change more strongly, certainly under Proposed Change Spec

Reply:
I added the following paragraph to the Proposed Change Specification:

> Permit nested foralls and contexts (a free order of quantifiers) in prefix-style constructor signatures (fix #18389).

And the following paragraph to Costs and Drawbacks:

> The free order of quantifiers in prefix-style constructors will have major effect on the implementation.